.changeset/new-rivers-fail.md

```diff
@@ -0,0 +1,54 @@
+---
+"@chakra-ui/toast": major
+"@chakra-ui/react": major
+---
+
+Add support for React 18.
+
+### Breaking changes for `createStandaloneToast`
+
+With React <17 Chakra UI was able to render the toast container DOM element for
+you. To allow Chakra UI to support all React versions >=16.8 you need to render
+the DOM element in your application code.
+
+#### before
+
+```ts
+import { createStandaloneToast } from "@chakra-ui/toast"
+
+const toast = createStandaloneToast()
+toast({ title: "Chakra UI" })
+```
+
+#### with React 17
+
+```tsx
+import { createStandaloneToast } from "@chakra-ui/toast"
+import * as ReactDOM from "react-dom"
+
+const toastContainerId = "chakra-toast-portal"
+const toastContainerElement = document.createElement("div")
+window.document.body.append(toastContainerElement)
+
+const { ToastContainer, toast } = createStandaloneToast()
+ReactDOM.render(<ToastContainer />, toastContainerElement)
+
+toast({ title: "Chakra UI" })
+```
+
+#### with React 18
+
+```tsx
+import { createStandaloneToast } from "@chakra-ui/toast"
+import * as ReactDOM from "react-dom/client"
+
+const toastContainerId = "chakra-toast-portal"
+const toastContainerElement = document.createElement("div")
+window.document.body.append(toastContainerElement)
+
+const { ToastContainer, toast } = createStandaloneToast()
+const root = ReactDOM.createRoot(toastContainerElement)
+root.render(<ToastContainer />)
```

Comment:
This change is targeting createStandaloneToast [see the current docs](https://chakra-ui.com/docs/components/feedback/toast#standalone-toasts) which allows you to create toasts from outside of your React root. E.g. some vanilla JS code. In this case chakra ui created the React root for you, but with the api change of React 18 I cannot think of a way how we could support both versions at the same time.

The most common toast function `useToast` did not change.

Reply:
I see, I had misread and thought that it would still inherit the portal from the provider, but I can see now that the `ToastContainer` renders its own standalone provider _and_ consumer.

One use case for `createStandaloneToast` currently is for error handling on a promise rejection within the context of a React application - e.g. failed fetch or error inside an event handler. Is that an "incorrect" usage for this?

The problem I see is when to mount/unmount the ToastContainer in the React lifecycle. If you call `createStandaloneToast` inside a callback then there is no cleanup method so you might as well just use the regular toast method - and perhaps that is the answer in this case.

Reply:
> One use case for createStandaloneToast currently is for error handling on a promise rejection within the context of a React application - e.g. failed fetch or error inside an event handler. Is that an "incorrect" usage for this?

The use case seems good to me!

I think it depends on where createStandaloneToast is called.

I call createStandaloneToast in its own file and reuse the toast function everywhere in my code where I don't have access to useToast. 

But if it's called in the React lifecycle it's best to cleanup the created Dom nodes.

Global state vs. cleanup 🤷

Reply:
If you mutate DOM in the global scope would that cause some weirdness during dev if you’re hot reloading the code?

Reply:
I’m really questioning the value of this helper function at all as part of a React library.

Reply:
> If you mutate DOM in the global scope would that cause some weirdness during dev if you’re hot reloading the code?

I am not an expert in hot reloading, but it haven't seen any Github issues regarding this.